File: js/json/src/snapshot/encoder.ts
import { Encoder as Flate } from "@moq/flate";
import type * as z from "zod/mini";

import { deepEqual, diff } from "../diff.ts";

// Maximum frames (snapshot + deltas) in a single group before a new snapshot is forced. Kept
// well below the per-group frame cap so a late joiner can always read the snapshot at frame 0.
const MAX_DELTA_FRAMES = 256;

// Delta ratio used when {@link Config.deltaRatio} is left unset. Not re-exported from the package
// entrypoint: the Producer needs it to know whether to hold a group open, nothing else does.
export const DEFAULT_DELTA_RATIO = 8;

/** Options shared by an {@link Encoder} and the {@link Producer} that wraps one. */
export interface Config<T> {
	// Controls how aggressively the encoder emits deltas (merge patches) instead of full snapshots.
	//
	// `0` disables deltas: every change is encoded as a new snapshot.
	//
	// A positive number enables deltas: a new snapshot is emitted once the deltas already written
	// to the current group (excluding the snapshot frame) exceed `deltaRatio` times the snapshot size.
	// The pending delta is excluded from that check, so the one that first crosses the budget still
	// lands before the group rolls. So `1` allows roughly one snapshot's worth of deltas before rolling.
	//
	// When {@link compression} is on, both sides of the comparison are measured on the compressed frame
	// sizes (the real wire cost).
	//
	// Defaults to `8` when unset.
	deltaRatio?: number;

	// Optional zod schema used to validate each value before publishing.
	schema?: z.ZodMiniType<T>;

	// Starting value for {@link Producer.mutate} before anything has been published. Required to
	// mutate a producer that hasn't published yet (e.g. a fresh catalog); ignored once a value exists.
	initial?: T;

	// Compress each group as one sync-flushed `deflate-raw` (RFC 1951) stream, so deltas reuse the
	// snapshot as context and shrink sharply. Interoperable with the Rust `moq-json` producer.
	// `false`/unset (the default) writes plaintext JSON frames. A {@link Decoder} reading them
	// must set the same flag.
	compression?: boolean;
}

/** One encoded frame, and the group boundary it implies. */
export interface Encoded {
	/** The frame payload, `deflate-raw` compressed when {@link Config.compression} is set. */
	payload: Uint8Array;

	/**
	 * Whether this frame is a full snapshot, which must open a new group.
	 *
	 * `true` means the caller writes it as the first frame of a fresh group; `false` means it is a
	 * merge patch that must be appended to the group the last snapshot opened.
	 *
	 * The encoder decides this, never the caller: a value that sets a field to JSON null, or whose
	 * root isn't an object, cannot be expressed as a merge patch at all, and the delta budget and
	 * frame cap force a snapshot independently of what the caller wanted.
	 */
	keyframe: boolean;
}

/**
 * Encodes a JSON value into frame payloads, choosing snapshots and deltas automatically.
 *
 * The track-free core of {@link Producer}: it decides *what bytes go in a frame* and *where the
 * group boundaries fall*, and leaves writing them to the caller. Reach for it when something else
 * already owns the track.
 *
 * Frames must reach the wire in the order they were encoded, and a frame with
 * {@link Encoded.keyframe} set must open a new group: both the merge patches and the group-scoped
 * DEFLATE window depend on it. If the caller closes a group for its own reasons, call
 * {@link reset} so the next value is encoded as a snapshot.
 */
export class Encoder<T> {
	#config: Config<T>;
	#compress: boolean;

	// The last encoded value, normalized through JSON so it matches what landed on the wire. The
	// baseline every delta is diffed against, and `undefined` until the first snapshot.
	#last?: unknown;

	// Bytes of deltas already emitted into the current group, excluding the snapshot frame. Compressed
	// frame sizes when compressing, raw otherwise, matching {@link #snapshotLen} so the budget check is
	// like-for-like (and identical to the Rust encoder).
	#deltaBytes = 0;
	// Size of the current group's snapshot frame, the reference the delta budget is measured against.
	#snapshotLen = 0;
	// Frames emitted into the current group, snapshot included.
	#groupFrames = 0;

	// The current group's `deflate-raw` stream, swapped for a fresh one (cold window) at each
	// snapshot, so a snapshot and its deltas share one stream.
	#flate?: Flate;

	constructor(config: Config<T> = {}) {
		this.#config = config;
		this.#compress = config.compression ?? false;
	}

	/** The last encoded value, or `undefined` before the first snapshot. */
	get value(): T | undefined {
		return this.#last as T | undefined;
	}

	/**
	 * Forget the baseline, so the next {@link update} emits a snapshot.
	 *
	 * Call this whenever the caller closes the current group behind the encoder's back. Without it
	 * the next value may be encoded as a delta against a DEFLATE window and a baseline that the new
	 * group doesn't carry.
	 */
	reset(): void {
		this.#last = undefined;
		this.#flate = undefined;
		this.#deltaBytes = 0;
		this.#snapshotLen = 0;
		this.#groupFrames = 0;
	}

	/**
	 * Encode a new value, as a snapshot or a delta.
	 *
	 * Returns `undefined` when the value is unchanged from the last one encoded, so nothing needs to
	 * be written.
	 */
	update(value: T): Encoded | undefined {
		const valid = this.#config.schema ? this.#config.schema.parse(value) : value;

		// Serialize once; parse it back to a normalized JSON value for diffing and comparison
		// (dropping `undefined` fields, matching what lands on the wire).
		const text = JSON.stringify(valid);
		const json = JSON.parse(text);
		if (this.#last !== undefined && deepEqual(this.#last, json)) return undefined;

		const delta = this.#delta(json);
		this.#last = json;

		if (delta) {
			const payload = this.#frame(delta);
			this.#deltaBytes += payload.length;
			this.#groupFrames += 1;
			return { payload, keyframe: false };
		}

		return { payload: this.#snapshot(new TextEncoder().encode(text)), keyframe: true };
	}

	// Resolved delta ratio: the configured value, or the default when unset. `0` disables deltas.
	get #deltaRatio(): number {
		return this.#config.deltaRatio ?? DEFAULT_DELTA_RATIO;
	}

	// Build a delta frame, or `undefined` to signal that a fresh snapshot should be emitted.
	//
	// The budget gate runs first, against the deltas already written, so rolling a new group costs no
	// merge-patch work. Since the gate excludes the frame about to be written, the delta that tips the
	// group past `ratio * snapshot` still lands: a group overshoots the budget by at most one delta.
	#delta(json: unknown): Uint8Array | undefined {
		const ratio = this.#deltaRatio;
		if (ratio === 0) return undefined;
		if (this.#last === undefined) return undefined;
		if (this.#groupFrames === 0 || this.#groupFrames >= MAX_DELTA_FRAMES) return undefined;

		// Gate on the deltas accumulated so far (snapshot frame excluded), before computing the patch.
		if (this.#deltaBytes > ratio * this.#snapshotLen) return undefined;

		const result = diff(this.#last, json);
		if (result.forcedSnapshot) return undefined;

		return new TextEncoder().encode(JSON.stringify(result.patch));
	}

	// Encode a group's snapshot (frame 0), returning its payload. On the compressed path this opens a
	// fresh stream (cold window), so the snapshot and its deltas share one DEFLATE window.
	#snapshot(frame: Uint8Array): Uint8Array {
		this.#flate = this.#compress ? new Flate() : undefined;

		const payload = this.#frame(frame);
		this.#snapshotLen = payload.length;
		this.#deltaBytes = 0;
		this.#groupFrames = 1;

		return payload;
	}

	// Compress a frame into the current group's window, or pass it through when uncompressed.
	#frame(frame: Uint8Array): Uint8Array {
		return this.#flate ? this.#flate.frame(frame) : frame;
	}
}

File: rs/moq-json/src/stream/encoder.rs
//! The track-free half of stream publishing: records in, frame payloads out.

use std::marker::PhantomData;

use bytes::Bytes;
use serde::Serialize;

use crate::{Error, Result};

/// Configuration for an [`Encoder`], and so for the [`Producer`](super::Producer) wrapping one.
///
/// Build from [`Default`] and override fields (the struct is `#[non_exhaustive]`, so new
/// options stay additive).
#[derive(Debug, Clone, Default)]
#[non_exhaustive]
pub struct ProducerConfig {
	/// Compress the group as one sync-flushed DEFLATE stream, so each record reuses the earlier
	/// ones as context and shrinks sharply.
	///
	/// `false` (the default) emits plaintext JSON frames. A [`Decoder`](super::Decoder) reading them
	/// must set [`ConsumerConfig::compression`](super::ConsumerConfig::compression) to match.
	pub compression: bool,
}

impl ProducerConfig {
	/// Set [`compression`](Self::compression) (a builder, since the struct is `#[non_exhaustive]`).
	pub fn with_compression(mut self, compression: bool) -> Self {
		self.compression = compression;
		self
	}
}

/// An encoded record the caller has not yet acknowledged writing.
///
/// Returned by [`Encoder::encode`]. Write the [`payload`](Self::payload), then
/// [`commit`](Self::commit).
///
/// A frame that is never committed never reached the wire. With compression on that is
/// unrecoverable within the group: the window is ahead of what the consumer holds, and a log has no
/// keyframe to resynchronize on the way [`snapshot`](crate::snapshot) does. So the encoder refuses
/// to encode anything further ([`Error::Desync`]) until the caller rolls a new group and calls
/// [`Encoder::reset`]. Without compression each record stands alone, so a dropped one leaves a gap
/// in the log but nothing undecodable, and encoding continues.
#[must_use = "the record must be written and committed, or the encoder will refuse to continue"]
pub struct Pending<'a, T> {
	encoder: &'a mut Encoder<T>,
	payload: Bytes,
	committed: bool,
}

impl<T> Pending<'_, T> {
	/// The frame payload to write.
	pub fn payload(&self) -> &Bytes {
		&self.payload
	}

	/// Acknowledge that the record reached the wire, keeping the encoder's window.
	///
	/// Only call this once the write has actually succeeded.
	pub fn commit(mut self) {
		self.committed = true;
	}
}

impl<T> Drop for Pending<'_, T> {
	fn drop(&mut self) {
		if !self.committed {
			self.encoder.desync();
		}
	}
}

/// Encodes JSON records into frame payloads, sharing one DEFLATE window across the log.
///
/// The track-free core of [`Producer`](super::Producer). Unlike
/// [`snapshot::Encoder`](crate::snapshot::Encoder) there are no group boundaries to report: a log is
/// an unbroken sequence of self-contained records, so every payload is simply the next frame.
///
/// The window spans everything encoded so far, so payloads must reach the wire in order and be
/// decoded in the same order. If the caller does roll a group, call [`reset`](Self::reset) so the
/// next record starts a cold window that the new group's decoder can follow.
pub struct Encoder<T> {
	/// The DEFLATE encoder (one window for the whole log), `Some` while compressing.
	flate: Option<moq_flate::Encoder>,
	compression: bool,

	/// Set when a compressed record was encoded but never written. The window is then ahead of the
	/// consumer for the rest of the group, so encoding stops until the caller rolls a new one.
	desynced: bool,

	_marker: PhantomData<fn(T)>,
}

impl<T> Encoder<T> {
	/// Create an encoder with a cold window.
	pub fn new(config: ProducerConfig) -> Self {
		Self {
			flate: config.compression.then(moq_flate::Encoder::new),
			compression: config.compression,
			desynced: false,
			_marker: PhantomData,
		}
	}

	/// Start a cold DEFLATE window, for a caller that has just rolled a group.
	///
	/// This is also how a caller clears an [`Error::Desync`]: roll a new group so the consumer starts
	/// its own cold window, then reset.
	pub fn reset(&mut self) {
		self.flate = self.compression.then(moq_flate::Encoder::new);
		self.desynced = false;
	}

	/// Mark the window as ahead of the consumer, after a record that was never written.
	///
	/// Only meaningful while compressing: an uncompressed record carries no shared state, so losing
	/// one leaves a gap in the log rather than an undecodable stream.
	fn desync(&mut self) {
		self.desynced = self.compression;
	}
}

impl<T: Serialize> Encoder<T> {
	/// Encode one record into the next frame payload.
	///
	/// The record comes back as a [`Pending`] the caller writes and then
	/// [`commit`](Pending::commit)s. Errors with [`Error::Desync`] if a previous compressed record
	/// was left uncommitted, since every frame after it would be undecodable.
	pub fn encode(&mut self, value: &T) -> Result<Pending<'_, T>> {
		if self.desynced {
			return Err(Error::Desync);
		}

		let bytes = serde_json::to_vec(value)?;
		let payload = match self.flate.as_mut() {
			Some(flate) => flate.frame(&bytes),
			None => Bytes::from(bytes),
		};

		Ok(Pending {
			encoder: self,
			payload,
			committed: false,
		})
	}
}
